bitcoin/src/bip152.rs

```diff
@@ -110,7 +109,7 @@ impl ShortId {
 
         // 2. Running SipHash-2-4 with the input being the transaction ID and the keys (k0/k1)
         // set to the first two little-endian 64-bit integers from the above hash, respectively.
-        (endian::slice_to_u64_le(&h[0..8]), endian::slice_to_u64_le(&h[8..16]))
+        (u64::from_le_bytes(h[0..8].try_into().expect("8 byte slice")), u64::from_le_bytes(h[8..16].try_into().expect("8 byte slice")))
```

Comment:
I think we could even make a *function* which took a `&[u8]` and promised to read exactly 2/4/8 bytes from it. It could even return a new slice starting right after the read, so that we could chain these calls (maybe .. I guess it has to return the actual LE number somehow too).

Reply:
Oh, yes, that makes sense. It may panic but that will be visible and documented. We kinda already have it:

```rust
use consensus::encode::ReadExt;
let mut hash_bytes = h.as_inner();
let first = hash_bytes.read_u64().unwrap();
let second = hash_bytes.read_u64().unwrap();
```

But could use some polishing.

Reply:
> I think we could even make a function which took a &[u8] and promised to read exactly 2/4/8 bytes from it.

At which time we are back exactly where we started with an `endian` module :) I'm not being facetious, after going around in circles on the `all::opcodes` PR I'm wondering if I'm not doing the same here on this PR. I'll have more of a play with it.

Reply:
Perhaps
```rust
        // 2. Running SipHash-2-4 with the input being the transaction ID and the keys (k0/k1)
        // set to the first two little-endian 64-bit integers from the above hash, respectively.

        // Verify at compile time that the two slices below won't cause a panic.
        bitcoin_internals::const_assert!(sha256::Hash::LEN >= 16);

        let mut buf = [0; 8];

        buf.copy_from_slice(h[0..8]);
        let first = u64::from_le_bytes(buf);

        buf.copy_from_slice(h[8..16]);
        let second = u64::from_le_bytes(buf);

        (first, second)
```

Reply:
I played around with doing the copy manually in various places as above and the idea of a helper function but the more I do the more I'm starting to prefer `let x = u32::from_be_bytes(data[9..13].try_into().expect("4 byte slice"));` for the following reasons
- The manual copy is not clean because it makes the function code operate at multiple layers of abstraction
- The `try_into` with clear `expect` string puts the panic and the invariant to prevent panic on the same line of code so its easy to check (in the example above one can check 9..13 does not breach the invariant described "4 byte slice" without jumping to another function definition.

Reply:
I agree, I prefer `.expect` to copying.

> At which time we are back exactly where we started with an endian module :) I'm not being facetious

Well, in fairness we'd have a simpler `endian` module which just wrapped the `expect` and was documented as "we're just wrapping `expect`s here". We could even call the module `slice` or `byteslice` and stick conversion functions for hashes to `secp256k1::Message` etc in there as well..

Reply:
I took a brief look at this and found there actually aren't that many places that subslice arrays and then convert to numbers or fixed-size arrays. So maybe just let it be as-is (or with additional fns) and return to it if we find more occurrences.

If you still want to do it I'd prefer statically checked API even if it's only implemented for some array sizes (up to 64 items or so). The impl won't be of course.

Reply:
Cool. I'm leaving it as is, the int-to-bytes bit is a clear improvement, the bytes-to-int is "standard" even though a bit messy with the `expect` calls, and the endian module is totally removed - so I'd call this a win still.